src/main/java/algebra/fields/Fp.java

```diff
@@ -102,9 +103,18 @@ public Fp multiplicativeGenerator() {
   }
 
   public Fp rootOfUnity(final long order) {
-    final BigInteger exponent = FpParameters.modulus().divide(BigInteger.valueOf(order));
-
-    return new Fp(FpParameters.root().modPow(exponent, FpParameters.modulus()), FpParameters);
+    final BigInteger root = FpParameters.root();
+    final long s = FpParameters.s();
+    final long logOrder = MathUtils.log2(order);
+    assert (order == 1 << logOrder);
+    assert (logOrder <= s);
+
+    // The integer root satisfies root^{2^s} = 1 mod r. Therefore, for order =
```

Comment:
`root^{2^s} = 1 mod r` -> let's use `mod p` instead to be consistent with the name of the class (i.e. `Fp`)

Reply:
replaced `r` with `p` in the comment.